File: MBA.R
getwd()
setwd("/Users/s.d.h.fernando/Documents/Level 04 Sem 01 /CS 4104/2025/Assignment 01")

#Load the libraries for each task
library(readxl)

# Load the dataset
retail_data <- read_excel("Online Retail.xlsx")

# View structure & summary
str(retail_data)
summary(retail_data)

# Check missing values by column
colSums(is.na(retail_data))

# Focus on CustomerID and Description
sum(is.na(retail_data$CustomerID))
sum(is.na(retail_data$Description))

#Load the library to handle missing values
library(dplyr)

# Remove rows with missing CustomerID or Description
retail_data <- retail_data %>%
  filter(!is.na(CustomerID), !is.na(Description))

# Check remaining rows after handling the missing values. 
nrow(retail_data) 

library(stringr)

# Remove rows where InvoiceNo starts with "C"
retail_data <- retail_data %>%
  filter(!str_starts(InvoiceNo, "C"))

# Check remaining rows
nrow(retail_data)

# Filter positive Quantity and UnitPrice
retail_data <- retail_data %>%
  filter(Quantity > 0, UnitPrice > 0)

# Check min values
summary(retail_data$Quantity)
summary(retail_data$UnitPrice)

# Drop InvoiceNo & CustomerID (not needed for Apriori)
retail_data <- retail_data %>%
  select(-CustomerID, -InvoiceDate, -StockCode)

# Keep cleaned dataset
head(retail_data)

write.csv(retail_data, "Cleaned_OnlineRetail.csv", row.names = FALSE)

library(ggplot2)

# Count transactions per country
country_counts <- retail_data %>%
  group_by(Country) %>%
  summarise(Total = n()) %>%
  arrange(desc(Total))

# View top countries
head(country_counts, 10)

# Plot top 10 countries
ggplot(head(country_counts, 10), aes(x = reorder(Country, -Total), y = Total)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  labs(title = "Top 10 Countries by Transactions", x = "Country", y = "Transaction Count") +
  theme_minimal()

#Quantity distribution
summary(retail_data$Quantity)
ggplot(retail_data, aes(x = Quantity))+
  geom_histogram(bins = 50, fill = "orange", color = "black") +
  xlim(0,100) +
  labs(title = "Distribution of Quantity", x = "Quantity", y = "Count")

#unit price distribution
summary(retail_data$UnitPrice)
ggplot(retail_data, aes(x = UnitPrice)) +
  geom_histogram(bins = 50, fill = "green", color = "black")+
  xlim (0,20) +
  labs(title = "Distribution of unit price", x = "Unit Price", y = "Count")

#Make the baskets 
# Top 3 countries by number of transactions
top_countries <- head(country_counts$Country, 3)
print(top_countries)

library(arules)
library(arulesViz)

#Apply Aprori algorithm

for (country in top_countries) {
  cat("\n====================", country, "====================\n")
  
  # Filter data for the country
  country_data <- retail_data %>% filter(Country == country)
  
  # Create transaction data: group products by InvoiceNo
  trans_data <- split(country_data$Description, country_data$InvoiceNo)
  
  # Convert to transactions
  trans <- as(trans_data, "transactions")
  
  # Summary of transactions
  cat("Number of transactions:", length(trans), "\n")
  inspect(trans[1:3])  # view first 3 transactions
  
  # Apply Apriori
  rules <- apriori(trans, parameter = list(supp = 0.01, conf = 0.5, minlen = 2))
  
  # Show top rules sorted by lift
  rules_sorted <- sort(rules, by="lift", decreasing=TRUE)
  inspect(head(rules_sorted, 10))
  
  # Plot rules
  plot(rules_sorted, method = "grouped")
}

File: MBA.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Load dataset
df = pd.read_excel("Online Retail.xlsx")

# --- Check structure & summary ---
print(df.info())
print(df.describe())
print(df.isnull().sum())

# --- Remove missing CustomerID or Description ---
df = df.dropna(subset=["CustomerID", "Description"])

# --- Remove cancellations (InvoiceNo starting with 'C') ---
df = df[~df["InvoiceNo"].astype(str).str.startswith("C")]

# --- Keep only positive Quantity & UnitPrice ---
df = df[(df["Quantity"] > 0) & (df["UnitPrice"] > 0)]

# --- Save cleaned dataset ---
df_clean = df.drop(columns=["CustomerID", "InvoiceDate", "StockCode"])
df_clean.to_csv("Cleaned_OnlineRetail.csv", index=False)

# --- Country distribution ---
country_counts = df_clean["Country"].value_counts()
print(country_counts.head(10))

sns.barplot(x=country_counts.head(10).index, y=country_counts.head(10).values)
plt.title("Top 10 Countries by Transactions")
plt.xticks(rotation=45)
plt.show()

# --- Quantity distribution ---
sns.histplot(df_clean["Quantity"], bins=50, color="orange")
plt.xlim(0,100)
plt.title("Quantity Distribution")
plt.show()

# --- Unit Price distribution ---
sns.histplot(df_clean["UnitPrice"], bins=50, color="green")
plt.xlim(0,20)
plt.title("Unit Price Distribution")
plt.show()

# --- Correlation ---
print(df_clean[["Quantity", "UnitPrice"]].corr())

from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder
import pandas as pd

# Assuming df_clean is your cleaned dataframe from before

top_countries = df_clean["Country"].value_counts().head(3).index
print("Top 3 Countries:", list(top_countries))

for country in top_countries:
    print(f"\n=== Apriori for {country} ===")

    # Filter data for the country
    country_data = df_clean[df_clean["Country"] == country]

    # Group items by InvoiceNo into baskets
    baskets = country_data.groupby("InvoiceNo")["Description"].apply(list).tolist()

    # Encode transactions
    te = TransactionEncoder()
    te_array = te.fit(baskets).transform(baskets)
    country_df = pd.DataFrame(te_array, columns=te.columns_)

    # Apply Apriori (support=1%, confidence=50%)
    frequent_itemsets = apriori(country_df, min_support=0.01, use_colnames=True)
    rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.5)

    # Sort rules by lift descending
    rules_sorted = rules.sort_values(by="lift", ascending=False)


    # Prepare rules for saving (convert frozensets to strings)
    def frozenset_to_string(x):
        return ', '.join(list(x))


    def clean_and_format_items(x):
        # Remove leading/trailing spaces and title case each item
        items = [item.strip().title() for item in x.split(',')]
        return ", ".join(items)


    # Apply the formatting function to antecedents and consequents after converting frozenset to string
    rules_sorted["antecedents"] = rules_sorted["antecedents"].apply(frozenset_to_string).apply(clean_and_format_items)
    rules_sorted["consequents"] = rules_sorted["consequents"].apply(frozenset_to_string).apply(clean_and_format_items)

    # Round numerical columns for better readability
    rules_sorted["support"] = rules_sorted["support"].round(4)
    rules_sorted["confidence"] = (rules_sorted["confidence"] * 100).round(4)  # in %
    rules_sorted["lift"] = rules_sorted["lift"].round(4)
    rules_sorted["leverage"] = rules_sorted["leverage"].round(4)
    rules_sorted["conviction"] = rules_sorted["conviction"].round(4)

    # Optionally convert confidence to a string with % symbol
    rules_sorted["confidence"] = rules_sorted["confidence"].astype(str) + "%"

    # Select relevant columns to save
    rules_to_save = rules_sorted[
        ["antecedents", "consequents", "support", "confidence", "lift", "leverage", "conviction"]]

    # Save to CSV
    filename = f"association_rules_{country.replace(' ', '_')}.csv"
    rules_to_save.to_csv(filename, index=False)

    print(f"Saved {len(rules_to_save)} rules to {filename}")







